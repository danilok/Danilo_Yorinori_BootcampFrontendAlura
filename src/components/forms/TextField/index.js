/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../foundation/Text';

const InputWrapper = styled.div`
  margin-bottom: 17px;
  margin-top: 12px;
`;

const Input = styled(Text)`
  width: 100%;
  border: 1px solid ${({ theme }) => theme.colors.tertiary.light.color};
  padding: 12px 16px;
  outline: 0;
  border-radius: ${({ theme }) => theme.borderRadius};
  border: 1px solid ${({ theme }) => theme.colors.background.dark.color};

  ${({ theme, isFieldInvalid }) => isFieldInvalid && css`
    border-color: ${theme.colors.danger.main.color};
    & + span {
      color: ${theme.colors.danger.main.color};
      font-size: 11px;
    }
  `}

  &:disabled {
    cursor: not-allowed;
    opacity: .5;
  }
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph1',
};

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  type,
  error,
  isTouched,
  ...props
}) {
  const hasError = Boolean(error);
  const isFieldInvalid = hasError && isTouched;
  return (
    <InputWrapper>
      <Input
        type={type}
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
        isFieldInvalid={isFieldInvalid}
        {...props}
      />

      {isFieldInvalid && (
        <Text
          variant="smallestException"
          color="danger.main"
          role="alert"
        >
          {error}
        </Text>
      )}
    </InputWrapper>
  );
}

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  type: PropTypes.string,
  error: PropTypes.string,
  isTouched: PropTypes.bool,
};

TextField.defaultProps = {
  type: 'text',
  error: '',
  isTouched: false,
};
