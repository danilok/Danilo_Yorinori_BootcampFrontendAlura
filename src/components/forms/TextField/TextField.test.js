import React from 'react';
import UserEvent from '@testing-library/user-event';
import { render, screen } from '../../../infra/test/testUtils';
import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Danilo"
        onChange={() => { }}
        name="nome"
      />,
    );
    const textField = screen.getByPlaceholderText(/nome/i);
    expect(textField).toMatchSnapshot();
  });

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', async () => {
        const user = UserEvent.setup();
        const onChangeMock = jest.fn();
        render(
          <TextField
            placeholder="Nome"
            value=""
            onChange={onChangeMock}
            name="nome"
            isTouched
          />,
        );
        const inputNome = screen.getByPlaceholderText(/nome/i);
        await user.type(inputNome, 'Danilo');
        expect(onChangeMock).toHaveBeenCalledTimes(6);
      });
    });
  });

  describe('when field is invalid', () => {
    test('displays the respective error message', () => {
      render(
        <TextField
          placeholder="Email"
          value="daniloemail.teste.com"
          onChange={() => { }}
          name="email"
          isTouched
          error="O campo email é obrigatório"
        />,
      );
      const inputEmail = screen.getByPlaceholderText(/email/i);
      expect(inputEmail).toHaveValue('daniloemail.teste.com');
      expect(screen.getByRole('alert')).toHaveTextContent('O campo email é obrigatório');
      expect(inputEmail).toMatchSnapshot();
    });
  });
});
