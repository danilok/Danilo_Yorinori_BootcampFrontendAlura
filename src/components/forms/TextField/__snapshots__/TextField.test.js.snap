// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-size: 18px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid #88989E;
  padding: 12px 16px;
  outline: 0;
  border-radius: 8px;
  border: 1px solid #004c40;
}

.c1:disabled {
  cursor: not-allowed;
  opacity: .5;
}

<input
  class="c0 c1"
  name="nome"
  placeholder="Nome"
  type="text"
  value="Danilo"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  font-size: 18px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid #88989E;
  padding: 12px 16px;
  outline: 0;
  border-radius: 8px;
  border: 1px solid #004c40;
  border-color: #dc1515;
}

.c1 + span {
  color: #dc1515;
  font-size: 11px;
}

.c1:disabled {
  cursor: not-allowed;
  opacity: .5;
}

<input
  class="c0 c1"
  name="email"
  placeholder="Email"
  type="text"
  value="daniloemail.teste.com"
/>
`;
